Fix author information leakage by author blocks for Custom Post Types without author support & display notice to user

sarthaknagoshe2002 opened the pull request:
fixes: #65816
## What?
This PR updates the behavior of author blocks to ensure they do not render when the post belongs to a Custom Post Type (CPT) that does not explicitly declare support for the author feature.

## Why?
Author blocks displaying on CPTs without declared author support is inconsistent with WordPress behavior.
This can lead to unintentional data exposure, as WordPress still saves author information to the database even when it is not intended to be visible.
Ensures better security, consistency, and adherence to user expectations for CPTs.

## How?

- Added a conditional check in the rendering logic of the author blocks to verify if the CPT supports the author feature.
- If the feature is unsupported, a message is displayed in the editor notifying users that the block will not render on the frontend.
- Updated the frontend output to prevent leaking author information when the CPT does not support authors.

## Testing Instructions
Reproduction:-
1. Create a Custom Post Type (CPT) without declaring support for the author feature:

```php
add_action( 'init', function() {
    register_post_type( 'custom-post', array(
        'public' => true,
        'supports' => array( 'title', 'editor', 'custom-fields' ),
    ) );
} );
```

2. Add a post to the newly created CPT. 
3. Add an author-related block (e.g., Post Author) in the block editor for that CPT post.
4. View the post on the frontend.

Expected Outcome:-

1. In the Editor:
    - A message is displayed informing users that the author block will not render because the CPT does not support the author feature.

2. On the Frontend:
    - The author block does not render, and no author information is displayed.

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/0b8bde8b-b032-44bc-b3da-03b04f8cc441


The suggestion given by @dhruvang21 is also used in this PR from the [comment](https://github.com/WordPress/gutenberg/issues/65816#issuecomment-2396080596)

sarthaknagoshe2002 committed:
fix: prevent author information from displaying author block for unsupported Custom Post Types & show notice to user

sarthaknagoshe2002 committed:
fix: PHP coding standards

Mamaduka commented:
Thanks for contributing, @sarthaknagoshe2002!

I'm unsure if we need to modify the block's edit components. People with access to the editors can also access details about post type author.

If it's decided to display a notice in edit components, it should probably be a part of rendered content and not an editor notice.

cc @WordPress/gutenberg-design

sarthaknagoshe2002 commented:
> If it's decided to display a notice in edit components, it should probably be a part of rendered content and not an editor notice.

@Mamaduka Can you elaborate more on this?

Mamaduka commented:
The blocks could render static text/warning in the editor when the post type doesn't support authors. But as I mentioned, I'm unsure if that warning makes sense in the editor context.

## Rought Example
![CleanShot 2024-11-20 at 19 21 46](https://github.com/user-attachments/assets/aeea79a6-6342-4674-9932-a665abe8d098)

sarthaknagoshe2002 committed:
fix: render static text warning instead of editor notice

sarthaknagoshe2002 commented:
@Mamaduka I have implemented your suggestion now.
In future if we conclude that the warning makes sense in the editor context then we can proceed with the merge.

## Screenshot
<img width="1470" alt="Screenshot 2024-11-21 at 12 34 10 AM" src="https://github.com/user-attachments/assets/bd7c41d3-c558-482a-9a0e-a41ace104f21">

sarthaknagoshe2002 commented:
I believe that the warning definitely makes sense in the editor context since displaying warnings/errors on the frontend doesn't seem appropriate.

Also, as you suggested static warning makes sense too.

sarthaknagoshe2002 commented:
@Mamaduka Any updates on how we should proceed with this PR?

Mamaduka commented on packages/block-library/src/post-author-name/edit.js:
```diff
@@ -33,6 +33,9 @@ function PostAuthorNameEdit( {
 
 			return {
 				authorName: _authorId ? getUser( _authorId ) : null,
+				supportsAuthor:
+					select( coreStore ).getPostType( postType )?.supports
+						?.author ?? false,
```
Let's extract `getPostType` above, alongside the `getUser`.

Mamaduka commented on packages/block-library/src/post-author-name/edit.js:
```diff
@@ -90,7 +93,15 @@ function PostAuthorNameEdit( {
 				</PanelBody>
 			</InspectorControls>
-			<div { ...blockProps }> { displayAuthor } </div>
+			{ supportsAuthor ? (
+				<div { ...blockProps }> { displayAuthor } </div>
+			) : (
+				<div { ...blockProps }>
+					{ __(
+						'The current post type does not support authors. The Post Author Name block will not be displayed.'
+					) }
+				</div>
+			) }
```
I think we can just do condition inside the `<div { ...blockProps }>`.

```js
{ supportsAuthor ? displayAuthor : __( 'The current post type does not support authors. The Post Author Name block will not be displayed.' ) }
```

Mamaduka commented on packages/block-library/src/post-author/edit.js:
```diff
@@ -52,6 +52,9 @@ function PostAuthorEdit( {
 				authorId: _authorId,
 				authorDetails: _authorId ? getUser( _authorId ) : null,
 				authors: getUsers( AUTHORS_QUERY ),
+				supportsAuthor:
+					select( coreStore ).getPostType( postType )?.supports
+						?.author ?? false,
```
Let's extract the `getPostType` above alongside the `getUser`.

Mamaduka commented on packages/block-library/src/post-author/edit.js:
```diff
@@ -189,55 +192,65 @@ function PostAuthorEdit( {
+					</div>
+				</div>
+			) : (
+				<div { ...blockProps }>
+					{ __(
+						'The current post type does not support authors. The Post Author block will not be displayed.'
```
Let's use early return for this instead for condition. It should be more readable.

You could do that just below the `showAuthorControl` variable definition.

Mamaduka reviewed (commented):
@sarthaknagoshe2002, I left a couple of suggestions. Still looking for design feedback before final approval.

P.S. There's also a proposal to deprecate Post Author block - #53427, but the information leaking issue also applies to deprecated blocks.

jameskoster commented:
> If it's decided to display a notice in edit components, it should probably be a part of rendered content and not an editor notice.

I agree with this, assuming you mean these inline warning notices (obviously with a different message):

<img width="1230" alt="Screenshot 2024-12-17 at 14 22 56" src="https://github.com/user-attachments/assets/4a0c08b5-cdf9-4cf3-8733-feab9961b672" />

Something like:

> This post type ($post_type) does not support Authors. [Remove block]

Additionally, should those blocks be hidden in the Inserter when editing a template associated with a post type that doesn't support them?

Mamaduka commented:
Thanks, @jameskoster!

Yes, the block editor warning component can work in this case.

> Additionally, should those blocks be hidden in the Inserter when editing a template associated with a post type that doesn't support them?

Generally, a template-aware inserter would be cool, but hiding blocks without context as to why they're hidden can be confusing.

sarthaknagoshe2002 committed:
Merge branch 'WordPress:trunk' into fix/issue-65816

fabiankaegy commented on packages/block-library/src/post-author-name/index.php:
```diff
@@ -26,6 +26,10 @@ function render_block_core_post_author_name( $attributes, $content, $block ) {
 		return '';
 	}
 
+	if ( ! post_type_supports( get_post_type( $block->context['postId'] ), 'author' ) ) {
```
```suggestion
	if ( ! post_type_supports( $block->context['postType'], 'author' ) ) {
```

We can directly get the post type via context as long as it's listed in the `usesContext` array in the `block.json`

sarthaknagoshe2002 committed:
Fix: optimize conditional rendering & address feedbacks

sarthaknagoshe2002 commented:
@fabiankaegy @Mamaduka I’ve addressed the feedback.  

@jameskoster I’ve updated the message to `This post type ($post_type) does not support Authors`. However, I’m unsure how to implement the `[Remove block]` functionality. Could you please guide me on this?

jameskoster commented:
I'm not actually sure myself, sorry. Maybe @Mamaduka knows 🤞

Mamaduka commented on packages/block-library/src/post-author/edit.js:
```diff
@@ -97,6 +99,18 @@ function PostAuthorEdit( {
+		return (
+			<div { ...blockProps }>
+				{ sprintf(
+					// Translators: %s is replaced with the name of the post type.
+					__( 'This post type (%s) does not support Authors.' ),
```
```suggestion
					// translators: %s: Name of the post type e.g: "post".
					__( 'This post type (%s) does not support the author.' ),
```

Mamaduka commented on packages/block-library/src/post-author-name/edit.js:
```diff
@@ -90,7 +93,17 @@ function PostAuthorNameEdit( {
+				{ supportsAuthor
+					? displayAuthor
+					: sprintf(
+							// Translators: %s is replaced with the name of the post type.
+							__(
+								'This post type (%s) does not support Authors.'
+							),
```
```suggestion
							// translators: %s: Name of the post type e.g: "post".
							__(
								'This post type (%s) does not support the author'
							),
```

fabiankaegy commented:
@sarthaknagoshe2002 you can remove a block like so:

```js
const BlockEdit = (props) => {
    const { clientId } = props;
    const { removeBlock } = useDispatch( blockEditorStore );
    
    return (
        <Button onClick={() => removeBlock(clientId)} isDestructive>Remove</Button>
    );
}
```

Mamaduka reviewed (commented):
I think there's no need to use the warning component and separate the removal action. Why?

* The severity of missing `author` support is negligible. It won't break or affect anything on the front end. You don't need the Waning component to grab the user's attention.
* Users can use toolbar options to remove the block.

@sarthaknagoshe2002 left minor suggestions regarding translation comments and string. After that, I think we're good to merge.

## Screenshot
![CleanShot 2024-12-18 at 18 21 54](https://github.com/user-attachments/assets/d13fd75f-8ae8-438d-bdb5-0bc121d1575e)

Mamaduka commented:
@fabiankaegy, the block editor component will receive the `onRemove` callback via props. But as I mentioned in my previous [comment](https://github.com/WordPress/gutenberg/pull/67136#pullrequestreview-2511977480), I think there's no need to implement it.

https://github.com/WordPress/gutenberg/blob/3159fa2864b9f1cd560a82cd81550eff1f66616a/packages/block-editor/src/components/block-list/block.js#L126

fabiankaegy commented:
> @fabiankaegy, the block editor component will receive the `onRemove` callback via props. But as I mentioned in my previous [comment](https://github.com/WordPress/gutenberg/pull/67136#pullrequestreview-2511977480), I think there's no need to implement it.
> 
> 
> 
> https://github.com/WordPress/gutenberg/blob/3159fa2864b9f1cd560a82cd81550eff1f66616a/packages/block-editor/src/components/block-list/block.js#L126

🤯 very good to know thank you :)

sarthaknagoshe2002 committed:
Fix: update the warning message & translator comment

sarthaknagoshe2002 commented:
@Mamaduka I’ve addressed the changes, and I believe this is now ready to be merged.

Mamaduka commented on packages/block-library/src/post-author-name/edit.js:
```diff
@@ -90,7 +93,17 @@ function PostAuthorNameEdit( {
+					: sprintf(
+							// translators: %s: Name of the post type e.g: "post".
+							__(
+								'This post type (%s) does not support the author'
```
```suggestion
								'This post type (%s) does not support the author.'
```

Mamaduka committed:
Missing period
